Implement DivRem intrinsic for X86

huoyaoyuan opened the pull request:
Closes #27292.

Some part of code was took from #37928 and #64864. I haven't fully understood all the concepts indeed. Needs some JIT expert to explain them.

Not implementing for Mono because I can't understand the code, and this one has special register constraints.

huoyaoyuan committed:
Add managed api for divrem

huoyaoyuan committed:
Add NI definition of DivRem

huoyaoyuan committed:
Fix DivRem to be static

huoyaoyuan committed:
Implement DivRem in clrjit

huoyaoyuan committed:
Add tests for DivRem

huoyaoyuan committed:
Adjust lsra and RMW

huoyaoyuan committed:
Use DivRem intrinsic in Math

huoyaoyuan committed:
Bring lower change from coreclr#37928
This fixes error while crossgen2 compiling Utf8Formatter.TryFormat(TimeSpan).

huoyaoyuan committed:
Fix signedness of DIV

huoyaoyuan committed:
Revert RMW change and fix reg allocation

huoyaoyuan committed:
Fix import of X64 intrinsic

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Math.cs:
```diff
@@ -317,13 +317,25 @@ public static int DivRem(int a, int b, out int result)
 
+            if (X86Base.IsSupported)
+            {
+                (int quitient, result) = X86Base.DivRem((uint)a, a >> 31, b);
```
There lacks a way to represent the `CDQ`/`CQO` instruction. Is there any way better than `>> 31`?

huoyaoyuan commented on src/tests/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx:
```diff
@@ -1261,6 +1261,20 @@ private static readonly (string templateFileName, Dictionary<string, string> tem
 
+private static readonly (string templateFileName, Dictionary<string, string> templateData)[] X86BaseInputs = new []
+{
+    ("ScalarTernOpTupleBinRetTest.template", new Dictionary<string, string> { ["Isa"] = "X86Base", ["Method"] = "DivRem", ["RetBaseType"] = "Int32",  ["Op1BaseType"] = "UInt32", ["Op2BaseType"] = "Int32",  ["Op3BaseType"] = "Int32",  ["NextValueOp1"] = "UInt32.MaxValue", ["NextValueOp2"] = "-2", ["NextValueOp3"] = "-0x10001",   ["ValidateResult"] = " int expectedQuotient = 0xFFFF;  int expectedRe ...
```
The tests were picked to verify correct signed-ness is used.

huoyaoyuan commented:
Code gen for sample method:
```csharp
        public static ulong XL(ulong a, ulong b)
        {
            (ulong q, ulong r) = Math.DivRem(a, b);
            return q + r;
        }
```

```asm
G_M52626_IG01:
       mov      qword ptr [rsp+10H], rdx
						;; bbWeight=1    PerfScore 1.00

G_M52626_IG02:
       xor      edx, edx
       mov      rax, rcx
       div      rdx:rax, qword ptr [rsp+10H]
       add      rax, rdx
						;; bbWeight=1    PerfScore 61.75

G_M52626_IG03:
       ret      
						;; bbWeight=1    PerfScore 1.00
; Total bytes of code: 19
```
In one commit it uses R8 instead of stack, but fails with register confliction in optimized code. 

It can be worse if `RDX` is used in calling convention. For the following method:
```csharp
        public long DivRemInt64(long a, long b)
        {
            (long q, long r) = Math.DivRem(a, b);
            return q + r;
        }
```

It saves a copy to stack:
```asm
       mov       [rsp+10],rdx
       sar       rdx,3F
       mov       rax,[rsp+10]
       idiv      r8
       add       rax,rdx
       ret
; Total bytes of code 21
```

And causes micro benchmark regresses. But real world sample can differ by register assignation.

huoyaoyuan committed:
Fix static in PNSE version

huoyaoyuan committed:
Fix accidential indent change

huoyaoyuan commented:
Do I need to implement for Mono in this PR or create an issue to track it? This is an already implemented ISA and used in corelib. Lack of implementation would cause existing tests to fail. /cc @vargaz

huoyaoyuan committed:
Apply format patch

huoyaoyuan committed:
op3 candidate should be different from op1 and op2

vargaz commented:
Maybe put the usage inside an #ifdef CORECLR for now and disable the relevant tests for mono ?

huoyaoyuan committed:
Add guard over AsHWIntrinsic

huoyaoyuan committed:
Disable DivRem intrinsic use for Mono

huoyaoyuan commented:
I can't prevent LSRA from allocating op3 into EAX, and it causes random assertion failures, no matter using `DelayFree` or not.
If I've understood correctly, both EAX and EDX are exposed by BuildDef, so there's no need to kill them right?

@dotnet/jit-contrib I need some help here.

huoyaoyuan committed:
Fix LSRA and invalid assertion

huoyaoyuan committed:
Set RWM information correctly, although totally unused.

huoyaoyuan commented:
All the XARCH tests passed. ARM64 failure looks like a timeout. Marking as ready.

huoyaoyuan committed:
Move multi-reg temp allocation to common helper

deeprobin reviewed (approved):
LGTM

huoyaoyuan commented:
The failure looks unrelated, but I'm not 100% sure that I didn't introduce any hole into JIT, since the usage of `Math.DivRem` is not rare.

huoyaoyuan commented:
I think I've understood how to implement it for Mono. However, currently Mono uses `X64` subclass to decide 64bit, which is unsuitable for `nint`. Leaving for the Mono team to implement it.

huoyaoyuan committed:
Disable DivRem intrinsic test for Mono.

EgorBo commented:
Nice! However, It feels weird to me to expose it as a public API - I think it's a trivial thing users expect JIT to handle like other compilers do.

EgorBo commented:
To clarify myself - it's fine and probably even great to have GT_INTRINSIC in JIT that this PR does so then we can recognize the div/mod pattern and replace with this intrinsic - I was only concerning about the C# API and the fact it's approved.

I'm looking forward to seeing it's merged

huoyaoyuan commented:
It's capable to do 128bit division, though maybe not useful.
The lack of CDQ/CQO method makes the potential result worse than recognizing the div then mod pattern. Is it reliable to recognize `>>31`?

tannergooding commented:
> Nice! However, It feels weird to me to expose it as a public API - I think it's a trivial (not in terms of this implementation, but in general) thing users expect JIT to handle like other compilers do

As from the API review, it is being exposed because x86/x64 `div`/`idiv` are special and are quite a bit more complex than just doing `div + rem`. They also allow `128 / 64`, `64 / 32`, `32 / 16`, and `16 / 8` operations as well as a few other specializations that make them "different enough" that it warrants a special API rather than a general xplat helper.

There are cases where the JIT itself could be improved here as well, certainly, but some of the specializations can be quite complex and this is something that even C/C++ compilers provide as part of their intrinsics repertoires.

danmoseley commented on src/libraries/System.Private.CoreLib/src/System/Math.cs:
```diff
@@ -317,13 +317,31 @@ public static int DivRem(int a, int b, out int result)
+#if !MONO
+            if (X86Base.IsSupported)
+            {
+                (int quitient, result) = X86Base.DivRem((uint)a, a >> 31, b);
```
typo "quotient" -- and below

huoyaoyuan replied:
The variable name actually conflicts with outer one. Using q and r for those.

huoyaoyuan committed:
Fix typo of quotient

tannergooding commented:
I can give this a general review, but its also going to require a review from someone in @dotnet/jit-contrib

tannergooding commented:
@huoyaoyuan, could you resolve the merge conflict and ensure this is up to date with the latest main? It's been a while and we need to ensure CI results are current.

This still needs review from @dotnet/jit-contrib

huoyaoyuan committed:
Merge branch 'main'

huoyaoyuan committed:
Merge branch 'main' into divrem

huoyaoyuan commented:
Ping for @dotnet/jit-contrib - can we get this in .NET 7?

huoyaoyuan committed:
Merge branch 'main'

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -22517,6 +22519,24 @@ GenTreeHWIntrinsic* Compiler::gtNewScalarHWIntrinsicNode(
                            /* isSimdAsHWIntrinsic */ false, op1, op2, op3);
 }
 
+GenTreeLclVar* Compiler::gtNewLclvForMultiRegIntrinsicNode(GenTreeHWIntrinsic* intrinsicNode, CORINFO_SIG_INFO* sig)
```
This seems duplicative with `impAssignMultiRegTypeToVar`.

huoyaoyuan replied:
That method is only defined under `FEATURE_MULTIREG_RET`, which is false under Windows x64. Multi-reg intrinsic wasn't treated as multi-reg return. It becomes quite mess then.

SingleAccretion replied:
You can just delete the `#ifdef`. Before this change, `FEATURE_MULTIREG_RET` was precisely equal to "supports multi-reg nodes in general".

SingleAccretion commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1800,27 +1800,15 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             //   it could point into the global heap or
             //   it could throw a null reference exception.
             //
             loadIntrinsic->gtFlags |= (GTF_GLOB_REF | GTF_EXCEPT);
```
I think you can delete this line, it should not be necessary now (but check we still set the flags in the constructor for the HWI node).

huoyaoyuan replied:
This looks somehow out of scope. I abstracted the common code path from arm64, but didn't meant to modify it. I don't have the environment to verify this change either.

huoyaoyuan replied:
`SetMultiReg` and `lvIsMultiRegRet` in fact doesn't do the same thing and replacing would cause failures.

kunalspathak replied:
`SetMultiReg()` goes on `GenTree` node while `lvIsMultiRegRet` goes on `LclVarDsc`. You missed setting the flag on `LclVarDsc`.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -384,6 +384,16 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
 
             ContainBlockStoreAddress(blkNode, size, dstAddr);
         }
+        else if (src->OperIsHWIntrinsic())
+        {
+            assert(!blkNode->AsObj()->GetLayout()->HasGCPtr());
+            if (blkNode->OperIs(GT_STORE_OBJ))
+            {
+                blkNode->SetOper(GT_STORE_BLK);
+            }
+            blkNode->gtBlkOpKind = GenTreeBlk::BlkOpKindUnroll;
+            ContainBlockStoreAddress(blkNode, size, dstAddr);
+        }
```
This does not look necessary.

The current design of multi-reg nodes is that they are always force-assigned to temps, i. e.:
```
multi-reg LCL_VAR = multi-reg node

Users should only see LCL_VAR
```
If we are seeing a multi-reg node on this path, this means something's not right upstream.

huoyaoyuan replied:
So what should be asserted here if we switch this into assertion?

SingleAccretion replied:
`assert(!src->IsMultiRegNode());` could be added at the top, but I don't think it's really necessary, things will naturally fail downstream if we see a multi-reg source here.

huoyaoyuan replied:
Let's remove this change and see if all the tests are fine.

SingleAccretion commented on src/coreclr/jit/lower.cpp:
```diff
@@ -6759,7 +6759,7 @@ bool Lowering::NodesAreEquivalentLeaves(GenTree* tree1, GenTree* tree2)
 {
     bool canEnregister = false;
-#if FEATURE_MULTIREG_RET
+#if FEATURE_MULTIREG_RET || FEATURE_HW_INTRINSICS
```
This does not look right, `FEATURE_HW_INTRINSICS` is a `#define` -- so presumably this should be `|| defined(FEATURE_HW_INTRINSICS)`.

kunalspathak commented on src/coreclr/jit/lower.cpp:
```diff
@@ -6776,10 +6776,10 @@ bool Lowering::CheckMultiRegLclVar(GenTreeLclVar* lclNode, const ReturnTypeDesc*
-    // For local stores on XARCH we only handle mismatched src/dest register count for
-    // calls of SIMD type. If the source was another lclVar similarly promoted, we would
+    // For local stores on XARCH we can't handle another lclVar source.
+    // If the source was another lclVar similarly promoted, we would
     // have broken it into multiple stores.
-    if (lclNode->OperIs(GT_STORE_LCL_VAR) && !lclNode->gtGetOp1()->OperIs(GT_CALL))
+    if (lclNode->OperIs(GT_STORE_LCL_VAR) && lclNode->gtGetOp1()->OperIs(GT_LCL_VAR))
```
Did you try to find the root cause of this?

huoyaoyuan replied:
I didn't have time for this. I brought it from another PR mentioned earlier and it just works. It's better to ask the original author.

huoyaoyuan replied:
@CarolEidt can you help explaining this? Not doing this isn't causing failures now, but the code looks necessary.

kunalspathak replied:
@huoyaoyuan - Carol doesn't work on this code base anymore. As the comment suggests, " If the source was another lclVar similarly promoted, we would have broken it into multiple stores.". So, I think you should revert this change, given that it works now.

huoyaoyuan replied:
It's better to provide an example involving this path so that I can learn from debugging. Currently, neither simple example nor the original failing `TryFormatDateTimeO` is triggering the caller method `LowerStoreLocCommon`.

huoyaoyuan replied:
It's not easy to debug this. It seems that `lvaEnregMultiRegVars` is only true under Release CLR.

Should it be changed to another check for `GT_HWINTRINSIC`, aside of `GT_CALL`?

SingleAccretion replied:
> It seems that `lvaEnregMultiRegVars` is only true under Release CLR.

(Are you running with optimizations on, aka TC off?)

> Should it be changed to another check for `GT_HWINTRINSIC`, aside of `GT_CALL`?

It's not clear that would be correct. Consider: `(uint, uint) a = DivRem(...)`, with the tuple on the LHS not promoted. I suspect it won't work right currently, with `a` enregistered as `long`. It is more of a problem with the check itself of course, as it only DNERs promoted locals.

Really, it seems the logic should be more like this:
```cs
if (comp->lvaEnregMultiRegVars && varDsc->lvPromoted) { ... }
else
{
    canEnreg = varTypeIsSIMD(lclNode);
#ifdef TARGET_XARCH
    // The current clause with the call.
#endif
}

if (canEnreg)
{
    lclNode->SetMultiReg();
}
else
{
    lclNode->ClearMultiReg();
    if (!varDsc->lvDoNotEnregister)
    {
        comp->lvaSetVarDoNotEnregister(lclNode->GetLclNum() DEBUGARG(DoNotEnregisterReason::BlockOp));
    }
}
```

huoyaoyuan replied:
The change is involving more and I have really no confidence to not break others. Who is the one with expertise in this?

kunalspathak replied:
I will spend some time to understand in more depth.

huoyaoyuan replied:
Reverting seems not valid - it's causing various different failures in CI.
But I have no idea why these cases are failing.

kunalspathak replied:
Yes, that was just a means to try which nodes are caught if `op1 != GT_CALL`. So the reason it is failing (as expected) is because the tree now contains the `op1 == GT_HWINTRINSIC`.

```
                                                            ┌──▌  t10    int    
                                                            ├──▌  t13    int    
                                                            ├──▌  t58    int    
N007 (  8,  9) [000014] -----------                   t14 = ▌  HWINTRINSIC struct int DivRem $140
                                                            ┌──▌  t14    struct 
N009 ( 18, 16) [000017] MA---------                         ▌  STORE_LCL_VAR struct<System.ValueTuple`2[System.Int32, System.Int32], 8>(P) V05 tmp3         
                                                            ▌    int    V05.Item1 (offs=0x00) -> V13 tmp11        
                                                            ▌    int    V05.Item2 (offs=0x04) -> V14 tmp12   
```

We do want to enregister in such cases and not doing it (retaining `op1 != GT_CALL`) spills the result on stack.

https://www.diffchecker.com/TUY5sGjV

SingleAccretion replied:
> Isn't this the same code on xarch?

Yes and no. My suggestion includes the (significant) addition that we will DNER _all_ locals, not just promoted ones. It is to prevent the case I described above from arising: `long = multi-reg-node` and other assigns codegen doesn't support. Alternatively, we could of course support them in codegen, but that seems out of scope for this change.

> Also, for the below code, do you mean to say use the current check of independent promotion too?

Yes, the independent promotion check (and the field count check as well) must stay. In fact, I think they should be expanded. Consider `struct { promoted field<double> } = HWI(int, int)` and other mismatched cases. Right now they would pass, but we need to DNER them, by passing the register count to the method instead of the return type descriptor.

huoyaoyuan replied:
There's now new failures and I don't have enough time to investigate in more depth.

kunalspathak replied:
I will take a look hopefully next week.

kunalspathak replied:
I updated the code as per @SingleAccretion suggestion.

kunalspathak commented:
> @kunalspathak, Libraries team needs a code review from JIT team. This is for .NET 7. PTAL.

I don't think we should merge this in .NET 7. There are many unknowns in the code and some, even I am not familiar with and having them right now feels risky. We can certainly revise and take it once we have a branch for .NET 8.

huoyaoyuan committed:
Merge branch 'main' into divrem

huoyaoyuan committed:
Use impAssignMultiRegTypeToVar

huoyaoyuan committed:
Update #ifdef

huoyaoyuan committed:
Remove change in LowerBlockStore

kunalspathak reviewed (commented):
Did a round of review and added some comments.

huoyaoyuan commented on src/coreclr/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1148,6 +1148,43 @@ void CodeGen::genX86BaseIntrinsic(GenTreeHWIntrinsic* node)
+            emitAttr attr = emitTypeSize(targetType);
+            emitter* emit = GetEmitter();
+
+            // op1: EAX, op2: EDX, op3: free
```
No constraint, can be reg or mem, confirmed with C++/asm.

kunalspathak replied:
In that case, just mention it that way.

tannergooding replied:
Looks like this is still pending getting a fix.

You might just say `reg/mem`

huoyaoyuan commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2193,6 +2209,26 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+                // DIV implicitly put op1(lower) to EAX and op2(upper) to EDX
+                srcCount += BuildOperandUses(op1, RBM_EAX);
+                srcCount += BuildOperandUses(op2, RBM_EDX);
+                srcCount += BuildOperandUses(op3);
```
> Do you have an example?

I did false assert when op3 is spilled onto stack, the regNum was remained as the register it was spilled from. This shouldn't be an issue now.

huoyaoyuan commented on src/coreclr/jit/compiler.h:
```diff
@@ -1871,10 +1871,10 @@ class Compiler
 #endif
 
-#if FEATURE_MULTIREG_RET
+#if FEATURE_MULTIREG_RET || defined(FEATURE_HW_INTRINSICS)
```
> but I never understand why this is `#if FEATURE_MULTIREG_RET ` and not `#ifdef FEATURE_MULTIREG_RET `

Because it's defined to 0/1.

> This seems to be set to 0 for windows/x64. It is slightly confusing to have `DivRem` that has multireg return on windows/x64 which doesn't support it?

`FEATURE_MULTIREG_RET` is about putting regular struct into multiple registers, while hw intrinsic is special and doesn't take part in regular ABI.

huoyaoyuan committed:
Adjust assert and helper method

huoyaoyuan committed:
Update CheckMultiRegLclVar

kunalspathak committed:
fix build errors

kunalspathak commented:
Failures related to https://github.com/dotnet/runtime/issues/76041

huoyaoyuan committed:
Merge branch 'main' into divrem

SingleAccretion commented on src/coreclr/jit/importer.cpp:
```diff
@@ -17043,19 +17043,25 @@ GenTree* Compiler::impAssignMultiRegTypeToVar(GenTree*             op,
 
     // Set "lvIsMultiRegRet" to block promotion under "!lvaEnregMultiRegVars".
-    lvaTable[tmpNum].lvIsMultiRegRet = true;
+    ret->SetMultiReg();
```
```suggestion
```
No longer needed (or expected).

JulieLeeMSFT commented:
@huoyaoyuan, please resolve the merge conflict.

kunalspathak commented:
> @huoyaoyuan, please resolve the merge conflict.

I will resolve this. As promised, I need to also investigate the CI errors that @huoyaoyuan was hitting.

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into pr-66551

kunalspathak commented:
I resolved merge conflicts and now failing to compile SPC.dll. On preliminary debugging looks like because of https://github.com/dotnet/runtime/pull/76766, we now expect the 2 structs involved in assignment should match, but they aren't. 

I hit this error:

```
D:\git\dotnet-runtime\src\coreclr\jit\morphblock.cpp:815
Assertion failed 'm_dst->OperIs(GT_LCL_VAR)' in 'System.Math:DivRem(ulong,ulong):System.ValueTuple`2[ulong,ulong]' during 'Morph - Global' (IL size 33; hash 0x81ec0472; FullOpts)
``` 

while dealing with the following:

```
               [000014] -A-XG------                         *  ASG       struct (copy)
               [000013] ---XG--N---                         +--*  BLK       struct<System.ValueTuple`2[System.UInt64, System.UInt64], 16>
               [000012] -----------                         |  \--*  LCL_VAR   byref  V00 RetBuf       
               [000011] m------N---                         \--*  LCL_VAR   struct<System.ValueTuple`2[System.UInt64, System.UInt64], 16>(P) V05 tmp1         
                                                            \--*    long   V05.Item1 (offs=0x00) -> V06 tmp2         
                                                            \--*    long   V05.Item2 (offs=0x08) -> V07 tmp3  
```

I will work with @SingleAccretion offline to understand more about this.

kunalspathak committed:
Fix some merge conflicts

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -23130,6 +23132,11 @@ ClassLayout* GenTreeHWIntrinsic::GetLayout(Compiler* compiler) const
     switch (GetHWIntrinsicId())
     {
+#ifdef TARGET_XARCH
+        case NI_X86Base_DivRem:
```
I think `NI_X86Base_DivRem` should get `typGetBlkLayout(8)`, since it is 8 bytes in size (being a `(uint, uint)` tuple).

tannergooding replied:
It can also be `16` in the case of `(nint, nint)`. We need to check the `simdBaseType` and handle it correspondingly -or- ensure that `nint` is always imported as `X64_DivRem` on 64-bit and as `_DivRem` on 32-bit.

kunalspathak replied:
> Seems like the code already correctly handles this with `genTypeSize(GetSimdBaseType()) * 2`.

Yep.

pentp commented on src/libraries/System.Private.CoreLib/src/System/Math.cs:
```diff
@@ -317,13 +317,31 @@ public static int DivRem(int a, int b, out int result)
+#if !MONO
+            if (X86Base.IsSupported)
+            {
+                (int quotient, result) = X86Base.DivRem((uint)a, a >> 31, b);
```
The signed divisions need to use `IDIV` instruction and can't use the unsigned `DIV`.

pentp replied:
Nvm, didn't notice that it will go to the correct overload based on the 2nd argument. But will the JIT be smart enough to use `CDQ/CDO` instructions here?
And constant divisors should not use the intrinsic at all.

kunalspathak replied:
> And constant divisors should not use the intrinsic at all.

Do you have a godbolt example to show what should happen ideally when you need both quotient and remainder for constant divisors?

pentp replied:
For constant divisors, the current code below is the ideal (or at least pretty close):
```cs
int div = a / b; // optimized by RyuJIT to a multiplication (+potentially some shifts)
result = a - (div * b);
return div;

kunalspathak replied:
So may be in JIT, we can check if the divisor is a constant, just do not detect it and proceed with fallback route.

kunalspathak replied:
> And CSE also probably doesn't work with this?

Correct, it doesn't CSE because of the following:

https://github.com/dotnet/runtime/blob/2aebbf75167b3aa21cae23e6fd702688dd724663/src/coreclr/jit/valuenum.cpp#L9522-L9531

@jakobbotsch

pentp replied:
Have you checked the codegen for callsites that use constant divisors for `Math.DivRem`? I don't see any mitigation for such cases, so I would NOT change `Math.DivRem` to use the new HW intrinsics in this PR.

kunalspathak replied:
```asm
G_M59029_IG02:
       mov      edx, ecx
       sar      edx, 31
       mov      r8d, 6
       mov      eax, ecx
       idiv     edx:eax, r8d
       mov      dword ptr [rsp], eax
       mov      dword ptr [rsp+04H], edx
       mov      rax, qword ptr [rsp]
```

pentp replied:
That's definitely worse and not what's expected by the users of the `Math.DivRem` API - so unless there's a good solution found for this, I'm still against using the intrinsic from `Math.DivRem`.

kunalspathak committed:
Fix some errors

kunalspathak committed:
Fix GCC build error

kunalspathak committed:
jit format

kunalspathak committed:
Make the size depending on the struct

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/X86Base.cs:
```diff
@@ -84,5 +97,27 @@ public static unsafe (int Eax, int Ebx, int Ecx, int Edx) CpuId(int functionId,
+        /// unsigned _udiv64(unsigned __int64 dividend, unsigned divisor, unsigned* remainder)
+        ///   DIV reg/m32
+        /// </summary>
+        public static (uint Quotient, uint Remainder) DivRem(uint lower, uint upper, uint divisor) => DivRem(lower, upper, divisor);
```
nit: Keeping these alphabetized helps with finding the right API in these otherwise very large files, so I'd expect it to appear above `Pause`

SingleAccretion commented on src/coreclr/jit/importer.cpp:
```diff
@@ -11308,16 +11308,19 @@ GenTree* Compiler::impAssignMultiRegTypeToVar(GenTree*             op,
+
+    LclVarDsc* varDsc = lvaGetDesc(tmpNum);
+
+    // The following is to exclude the fields of the local to have SSA.
```
```suggestion
    // Set "lvIsMultiRegRet" to block promotion under "!lvaEnregMultiRegVars".
```
(The SSA bit is no longer true)

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
Fix the formatting of summary

kunalspathak committed:
Remove trailing spaces

kunalspathak commented:
Fresh assert:

```
Assert failure(PID 5880 [0x000016f8], Thread: 1408 [0x0580]): Assertion failed 'idx < MAX_RET_REG_COUNT' in 'System.Version:TryFormat(System.Span`1[ushort],int,byref):bool:this' during 'LSRA allocate' (IL size 245; hash 0x69e91535; Tier1)

    File: D:\a\_work\1\s\src\coreclr\jit\gentree.h Line: 3747
    Image: C:\h\w\B09B0A05\p\dotnet.exe
```

kunalspathak committed:
Fix the tests to adopt new model

kunalspathak committed:
Fix MAX_RET_REG_COUNT -> MAX_MULTIREG_COUNT

kunalspathak commented:
@akoeplinger - who is the best person to talk to about the failures in `DivRem`?

kunalspathak committed:
Exclude from mono run

kunalspathak commented:
> @akoeplinger - who is the best person to talk to about the failures in `DivRem`?

never mind. It was related to https://github.com/dotnet/runtime/issues/75767. I added annotations to skip for mono.

kunalspathak committed:
Add RequiresProcessIsolation

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
Real fix for build break

kunalspathak committed:
Fix the test cases to adopt to the new system

kunalspathak committed:
Disable for llvm-fullaot

kunalspathak committed:
Also continue disabling for mono

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
Disable the test in a different group

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
fix merge conflict

kunalspathak committed:
format

kunalspathak committed:
fix another merge conflict error

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into divrem

kunalspathak committed:
misc changes from review

kunalspathak committed:
fix the replay errors

kunalspathak committed:
jit format

kunalspathak committed:
Add exclude list in mono/llvmfullaot

pentp reviewed (changes requested):
Unconditionally using the intrinsic from `Math.DivRem` API is going to cause a lot of regressions for cases with constant divisors (quite common in the runtime itself also).
IMHO this PR should be limited to the intrinsic itself and not change `Math` APIs.

kunalspathak commented:
I agree. I checked the current implementation vs. `DivRem` and with `DivRem`, it definitely looks worse. `MyTest1` is the current implementation while `MyTest2` is with `DivRem`. Same with `MyTest_const1` and `MyTest_const2`.

[long_example.txt](https://github.com/dotnet/runtime/files/10464007/long_example.txt)
[int_example.txt](https://github.com/dotnet/runtime/files/10464008/int_example.txt)
[sbyte_example.txt](https://github.com/dotnet/runtime/files/10464010/sbyte_example.txt)
[int_example_return_tuple.txt](https://github.com/dotnet/runtime/files/10464043/int_example_return_tuple.txt)

kunalspathak commented on src/coreclr/jit/lower.cpp:
```diff
@@ -6900,7 +6900,8 @@ bool Lowering::NodesAreEquivalentLeaves(GenTree* tree1, GenTree* tree2)
 bool Lowering::CheckMultiRegLclVar(GenTreeLclVar* lclNode, const ReturnTypeDesc* retTypeDesc)
```
I fixed the logic of `CheckMultiRegLclVar()` the way you suggested. I am not fully sure about this example though. Can you give C# example where I can see things goes wrong?

SingleAccretion replied:
> Can you give C# example where I can see things goes wrong?

It is not unlikely this cannot actually be reproduced right now because of the constrained way multi-reg locals are used, but here would be the idea for a reproduction:
```
// V00: struct { long, long }
// V01: struct { int, int, int, int }

V00 = DivDem()
*(typeof(V00)*)V01 = V00
```
Then V00 and V01 need to be promoted and forward subtitution needs to happen s.t. we end with:
```
V01 = DivRem()
```
The more general point though is that this _can_ happen according the IR validity rules and so needs to be handled.

kunalspathak committed:
Review feedback and fix bug in CheckMultiRegLclVar

pentp commented:
> > Unconditionally using the intrinsic from Math.DivRem API is going to cause a lot of regressions
> 
> I agree. I checked the current implementation vs. `DivRem` and with `DivRem`, it definitely looks worse. `MyTest1` is the current implementation while `MyTest2` is with `DivRem`. Same with `MyTest_const1` and `MyTest_const2`.

From these samples it looks like the new codegen is actually worse for the non-constant case also? It's repeatedly copying stuff to/from memory instead of using registers?

kunalspathak commented:
@akoeplinger - can you confirm how do I disable these tests for mono linux x64? I tried various ways but they do not seem to work.

kunalspathak committed:
Pass registerCount

SingleAccretion commented on src/coreclr/jit/lower.cpp:
```diff
@@ -3448,7 +3448,7 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
         if (comp->compMethodReturnsMultiRegRetType() && retVal->OperIs(GT_LCL_VAR))
         {
-            CheckMultiRegLclVar(retVal->AsLclVar(), &comp->compRetTypeDesc);
+            CheckMultiRegLclVar(retVal->AsLclVar(), retVal->GetMultiRegCount(comp));
```
```suggestion
            CheckMultiRegLclVar(retVal->AsLclVar(), comp->compRetTypeDesc.GetReturnRegCount());
```
The register count is for how the local will be used in this case.

SingleAccretion commented on src/coreclr/jit/lower.cpp:
```diff
@@ -6890,15 +6884,9 @@ bool Lowering::NodesAreEquivalentLeaves(GenTree* tree1, GenTree* tree2)
 //   lclNode     - the GT_LCL_VAR or GT_STORE_LCL_VAR node.
-//   retTypeDesc - a return type descriptor either for a call source of a store of
-//                 the local, or for the GT_RETURN consumer of the local.
+//   registerCount - register count of source of a store.
```
```suggestion
//   registerCount - use register count for uses; source register count for stores
```

kunalspathak committed:
review feedback

pentp commented:
> Actually, I think my binaries were out of date when I generated them. Here are the latest code and they look better. I am inclined to add back the intrinsic usage in `Math.DivRem`. Let me know what you think.

The codegen for non-const divisor now looks more like expected, but there's still a memory copy for the divisor and manual dividend extensions instead of using `CQO`. Currently there's only a tiny potential improvement with significant risk for regressions (code size from 22 to 24 and PerfScore from 75.45 to 74.40 for `long`).
It's probably relatively hard to achieve, but ideally the codegen would be:
```diff
G_M19802_IG01:
-       mov      qword ptr [rsp+10H], rdx
+       mov      r9, rdx 
						;; size=5 bbWeight=1 PerfScore 1.00
G_M19802_IG02:
-       mov      rdx, rcx
-       sar      rdx, 63
       mov      rax, rcx
+       cqo      
-       idiv     rdx:rax, qword ptr [rsp+10H]
+       idiv     rdx:rax, r9
       mov      qword ptr [r8], rdx
						;; size=18 bbWeight=1 PerfScore 70.00
G_M19802_IG03:
       ret
```

The constant divisor case still looks non-usable - there's a code size improvement from 50 to 33, but PerfScore goes from 14.75 to 76.55 because it uses `IDIV`.

kunalspathak commented:
Test failures are https://github.com/dotnet/runtime/issues/74838

kunalspathak committed:
Remove the extra comments

kunalspathak commented:
Mono linux x64 keeps failing despite disabling it. @akoeplinger will look into it tomorrow. But apart from that, I think this is in good shape. @tannergooding  - can you take another look?

kunalspathak commented:
/azp run runtime-coreclr superpmi-replay

akoeplinger committed:
Disable tests on mono llvmfullaot runs as well

kunalspathak committed:
Add missing case for upper save/restore

kunalspathak committed:
Add missing RequiresProcessIsolation

kunalspathak committed:
Revert "Add missing case for upper save/restore"

This reverts commit a98dbde4bd547824388fcb410c06b2503a870fb6.

kunalspathak committed:
Merge remote-tracking branch 'huoyaoyuan/divrem' into divrem

tannergooding commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2152,6 +2168,26 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+                // DIV implicitly put op1(lower) to EAX and op2(upper) to EDX
+                srcCount += BuildOperandUses(op1, RBM_EAX);
+                srcCount += BuildOperandUses(op2, RBM_EDX);
+                srcCount += op3->isContained() ? BuildOperandUses(op3) : BuildDelayFreeUses(op3, op1);
```
Is this going to work as expected?

Normally the second operand to `BuildDelayFreeUses` is the rmw node. However, in this case we have two RMW nodes (op1 and op2) and so we should likely need to indicate that `op3` is delay free with regards to both `op1` and `op2`.

kunalspathak replied:
I believe it remains `delayRegFree` for the entire treeNode. We reset `pendingDelayFree` when processing the next `treeNode`.

tannergooding replied:
Hmmm, @kunalspathak, there is special logic that looks at `node` (`op3`) compared to `rmwNode` (`op1`)

and a note specifically around:
```cpp
// Note: we don't handle multi-reg vars here. It's not clear that there are any cases
// where we'd encounter a multi-reg var in an RMW context.
assert(!rmwNode->AsLclVar()->IsMultiReg());
```

So my guess is we either don't have sufficient tests covering this case or we aren't correctly tracking the return value as multi-reg. There is also has the later logic that says `if (node == rmwNode)` then it might not set the delay free.

So, I think we need another overload of `BuildDelayFreeUses` that takes both `rmwNodes` and does the right general checks.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/X86Base.cs:
```diff
@@ -44,6 +45,18 @@ public abstract class X64
+            /// unsigned __int64 _udiv128(unsigned __int64 highdividend, unsigned __int64 lowdividend, unsigned __int64 divisor, unsigned __int64* remainder)
+            ///   DIV reg/m64
+            /// </summary>
+            public static (ulong Quotient, ulong Remainder) DivRem(ulong lower, ulong upper, ulong divisor) => DivRem(lower, upper, divisor);
```
I'd expect a few managed methods, such as `Int32/Int64/Math/UInt32/UInt64.DivRem` to be updated to use this handling or be likewise directly imported as this intrinsic (as part of this PR).

Long term I'd expect we also add an opt to `morph` to recognize `/` + `%` and optimize down to `DivRem`, but that's likely out of scope of this PR.

kunalspathak replied:
I had a discussion with @pentp in https://github.com/dotnet/runtime/pull/66551#issuecomment-1398928443 and offline and we decided to not do it in this PR.

tannergooding replied:
I think it's better to go ahead and utilize this anyways, particularly for the `Int32/Int64/UInt32/UInt64` cases that have identical signatures. That will help improve test coverage, increase confidence in correctness, and can be used to help drive fixing the other issue if it's a big enough of an impact.

kunalspathak replied:
> That will help improve test coverage, increase confidence in correctness, and can be used to help drive fixing the other issue 

That is definitely an advantage. @pentp ?

pentp replied:
I'm also in favour of using it for `Math.DivRem` and similar, but that would require solving some additional issues that we felt would be best addressed in a separate follow-up PR. Namely the constant divisor case would need to utilize `RuntimeHelpers.IsKnownConstant` to bypass the intrinsic (and new overloads need to be added for `IsKnownConstant`) and the regular divisor cases had some occasional register allocation issues also that caused unnecessary memory copies. Plus there's the issue of forfeiting use of `CDQ/CQO` that's probably not easily solved at C# level and would instead need some special handling at the importer (or something like that).

So I would like to see the intrinsic itself merged and then get the follow-up use cases as smaller PRs.

tannergooding replied:
What is tracking this work being done as part of .NET 8

I don't think we can/should check in the API if it's being exposed in such a way that it is worse than public API for the typical scenario. That will lead developers down the wrong path for an API that is meant to be oriented towards performance scenarios.

Based on the above, most of these look like trivial 1-2 cycle differences that may end up balancing out for other reasons. I don't see anything that indicates the cost increase is significant or should block its broader use.

Making it more used now will also cause any necessary additional work, such as what appear to be constant folding support and basic containment of casts, to be prioritized as part of .NET 8.
-------------

> Namely the constant divisor case would need to utilize RuntimeHelpers.IsKnownConstant to bypass the intrinsic (and new overloads need to be added for IsKnownConstant)

I don't think this is needed. We just need to ensure the relevant constant folding support exists for `NI_X86Base_DivRem` instead. `IsKnownConstant` has known issues/limitations and increases the size of the imported IR which can have other negative consequences/limits.

> the regular divisor cases had some occasional register allocation issues that caused unnecessary memory copies.

This doesn't seem like a blocker and is something that any code can hit. It sounds like something we can either live with or push towards trivially fixing separately. At a glance, I believe that this can be trivially handled with a couple tweaks in codegen and the lsra building for this intrinsic.

> Plus there's the issue of forfeiting use of CDQ/CQO that's probably not easily solved at C# level and would instead need some special handling at the importer (or something like that).

This should really just be a special case of processing (containing) `CAST` as part of the HWIntrinsic, it should likewise not be a blocker to the feature being merged and used more generally.

huoyaoyuan replied:
> I'd expect a few managed methods, such as `Int32/Int64/Math/UInt32/UInt64.DivRem` to be updated to use this handling or be likewise directly imported as this intrinsic (as part of this PR).

I had a prototype branch on my another machine, for `Decimal`. I planned to put off it after this PR.

kunalspathak committed:
Remove RequiresProcessIsolation property

kunalspathak commented:
I am currently waiting to get the tests working for https://github.com/dotnet/runtime/pull/80297 because it too has same problem we want to solve here. We don't want the new APIs to get consumed publically just yet, so need to devise a way such that test projects should be still able to consume them.

kunalspathak committed:
Merge branch 'main' into divrem

kunalspathak committed:
Replace ref with fakelibs

kunalspathak committed:
Add references of fakelib to test project

kunalspathak committed:
Fix CompatibilitySuppressions.xml

kunalspathak committed:
fix test builds

kunalspathak committed:
Remove unneeded extern/using

vargaz commented:
The mono failure is:
```
EXEC : error : Loaded assembly '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll' doesn't match original file name '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll'. Set MONO_PATH to the assembly's location. [/__w/1/s/src/mono/msbuild/aot-compile.proj]
/__w/1/s/src/mono/msbuild/aot-compile.proj(19,9): error MSB3073: The command "/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/corerun /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll" exited with code 1.
```
When running the aot compiler, the MONO_PATH env var needs to be set in such a way that the input corlib assembly is the same assembly as the one found in the directories in MONO_PATH. Not sure why this only fails for this one test.

kunalspathak commented:
>  Not sure why this only fails for this one test.

Is there a way to disable these tests?

kunalspathak committed:
Revert "Replace ref with fakelibs"

This reverts commit e1b5fb38ce84c34f143e36be6aafc27ce820d8df.

kunalspathak committed:
Revert "Revert "Replace ref with fakelibs""

This reverts commit 39ae98d2fee0c2b9d94437bce744c9f065654ed0.

kunalspathak committed:
Exclude DivRem.csproj

kunalspathak committed:
Handle the case to delay-free op3 for op1/op2

kunalspathak commented:
> When running the aot compiler, the MONO_PATH env var needs to be set in such a way that the input corlib assembly is the same assembly as the one found in the directories in MONO_PATH. Not sure why this only fails for this one test.

I do see that this path is setup correctly.

```
2023-02-13T18:51:19.8150781Z   aot-compile: compiling /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll; MONO_PATH: /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r:/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root
2023-02-13T18:51:19.8580478Z EXEC : error : Loaded assembly '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll' doesn't match original file name '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll'. Set MONO_PATH to the assembly's location. [/__w/1/s/src/mono/msbuild/aot-compile.proj]
```

@lambdageek

lambdageek commented:
> > When running the aot compiler, the MONO_PATH env var needs to be set in such a way that the input corlib assembly is the same assembly as the one found in the directories in MONO_PATH. Not sure why this only fails for this one test.
> 
> I do see that this path is setup correctly.
> 
> ```
> 2023-02-13T18:51:19.8150781Z   aot-compile: compiling /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll; MONO_PATH: /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r:/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root
> 2023-02-13T18:51:19.8580478Z EXEC : error : Loaded assembly '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/System.Private.CoreLib.dll' doesn't match original file name '/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/HardwareIntrinsics/HardwareIntrinsics_r/System.Private.CoreLib.dll'. Set MONO_PATH to the assembly's location. [/__w/1/s/src/mono/msbuild/aot-compile.proj]
> ```
> 
> @lambdageek

The issue is there's two different libraries in the MONO_PATH both called CoreLib.dll - the "real" one and the fake one that the testsuite adds in order to allow the tests to call APIs that are nto in System.Runtime.

That is very likely to confuse the AOT compiler.  It would be much better if the fake one didn't end up in the directory for the test case.

I think it's worth investigating if the fake one can be a reference assembly, instead.  But I'm not sure even that would be enough.

I'll try to build this PR (or #80297, which uses the same trick) locally to see if there's some way to make the AOT compiler happy

lambdageek commented:
Left a comment on the other PR that does the same fake corelib trick with one solution: https://github.com/dotnet/runtime/pull/80297#issuecomment-1430415392

Unfortunately that solution just exposed an issue where the new intrinsic created some condition that the AOT compiler did not expect to see.

But maybe it'll work on this PR. Here's the commit (for the other PR) that shows the approach https://github.com/lambdageek/runtime/commit/347d9437162858f2c48237547a3ed9b810ef2f5a

kunalspathak committed:
Create the DivRem.RefOnly fake CoreLib as a reference assembly

Make the DivRem tests reference the DivRem fake
CoreLib as reference assembly; and ensure that it is not included as a
ProjectReference by the toplevel HardwareIntrinsics merged test
runners.

The upshot is that the DivRem tests can call the extra APIs via
a direct reference to CoreLib (instead of through System.Runtime), but
the fake library is not copied into any test artifact directories, and
the Mono AOT compiler never sees it.

tannergooding commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -3256,24 +3256,66 @@ void LinearScan::setDelayFree(RefPosition* use)
+//    rmwNode           - The node that has RMW semantics (if applicable)
+//    useRefPosition - The use refposition that need to be delay-freed.
+//
+void LinearScan::DelayFreeUses(RefPosition* useRefPosition, GenTree* rmwNode)
```
Should this be called something like `AddDelayFreeUses` and called from `BuildDelayFreeUses` to ensure future refactorings/changes don't update one place but not the other?

kunalspathak replied:
Sure, done.

tannergooding commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2178,7 +2178,26 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+
+                    RefPosition* op3RefPosition;
+                    srcCount += BuildDelayFreeUses(op3, op1, RBM_NONE, &op3RefPosition);
+                    if ((op3RefPosition != nullptr) && !op3RefPosition->delayRegFree)
```
We don't need to also set it as delay free for `op3` because we don't track `delayFree` with regards to a specific operand, is that right?

kunalspathak replied:
Correct. We just want to make sure that `op3` is marked as `delayFree` so neither of `op1` or `op2`'s registers conflict with that of `op3`.

kunalspathak committed:
Unify AddDelayFreeUses

tannergooding reviewed (approved):
Would be good to have tracking issues covering the mono work and any work required to use this from Math.DivRem before merging.

kunalspathak commented:
> Would be good to have tracking issues covering the mono work and any work required to use this from Math.DivRem before merging.

Created:
- https://github.com/dotnet/runtime/issues/82194
- https://github.com/dotnet/runtime/issues/82195

jkotas commented on src/libraries/System.Runtime.Intrinsics/src/CompatibilitySuppressions.xml:
```diff
@@ -0,0 +1,76 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!-- https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids -->
+<Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
```
Why do we have all these compatibility suppressions instead of adding these API to ref assemblies?

jkotas commented on src/tests/JIT/HardwareIntrinsics/X86/X86Base/DivRem.RefOnly.csproj:
```diff
@@ -0,0 +1,22 @@
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <OutputType>Library</OutputType>
+    <CLRTestKind>SharedLibrary</CLRTestKind>
+    <AssemblyName>System.Private.CoreLib</AssemblyName>
```
Referencing CoreLib from the tests is anti-pattern. Tests should only reference public surface. If the tests need to hack on internal implementation details, they should use private reflection. Why are we doing this?

jkotas commented:
The structure of introduce by this PR is very non-standard and full of anti-patterns. I am tempted to revert it so that it can be done properly. @ViktorHofer @akoeplinger Thoughts?

jkotas commented:
> We don't want the new APIs to get consumed publically just yet, 

Why not? 

If you are worried that the Mono implementation is not in place yet, it would be fine to add a temporary non-intrinsic implementation under MONO ifdef. It would be very local change and it would avoid spreading the CoreLib hacks through many places.

kunalspathak commented:
I agree and I too do not like these hacks. Unfortunately, we had to go this route because there are few things, we need to optimize in order to consume it in `Math.DivRem` (Related https://github.com/dotnet/runtime/pull/66551#discussion_r1087392606). We wanted to get in this work, but not expose it just yet and so @tannergooding , @pentp  and me agreed to go this route. This is not even about mono at this point but also applicable to coreclr. If you think we should revert it, we can revert it now, fix the optimizations that would make `DivRem` ready and consumable (e.g. `Math.DivRem`) and re-merge this one in?

kunalspathak commented:
We also had planned to do similar thing in https://github.com/dotnet/runtime/pull/80297 since the new APIs that I am adding are not yet approved, but I decided to rather wait for API approval than to merge such anti-patterns of consuming them in test projects.

jkotas commented:
It is common to add new public APIs with simpler implementation with less-than-ideal performance first, and then optimize the implementation in subsequent PRs. We have done that many times. For example, #77799 added initial implementation of FrozenColllections and then number of subsequent PRs optimized it further, and I am sure we will get some more optimizations before .NET 8 ships. We should follow the same pattern for codegen intrinsics.

If you are really worried about the API being used accidentally before it is ready for prime time, you can slap `RequiresPreviewFeaturesAttribute` on it, like how it is done for `AvxVnni` Intrinsics currently.

jkotas commented:
> We also had planned to do similar thing in https://github.com/dotnet/runtime/pull/80297 since the new APIs that I am adding are not yet approved, but I decided to rather wait for API approval than to merge such anti-patterns of consuming them in test projects.

Yes, please do wait for the API shape to be approved first before merging the implementation.

kunalspathak commented:
> If you are really worried about the API being used accidentally before it is ready for prime time, you can slap RequiresPreviewFeaturesAttribute on it, like how it is done for AvxVnni Intrinsics currently.

I see. Ok, I will submit a PR to revert part of the hacks and add `RequiresPreviewFeaturesAttribute`.

tannergooding commented:
Just noting these are hardware intrinsics and providing a software fallback is itself problematic and against the general design goal of the APIs and can lead users into a pit of failure. If one is provided, it needs to very explicitly still throw where the underlying platform is not x86/x64 and when `X86Base.IsSupported` reports `false`.

That being said, exposing a platform specific intrinsic API publicly without it being usable for its intended purpose is likewise problematic and leads users into a different pit of failure. This is why I originally pushed for the general support to be completed and for the API to be used from the primary consumption sites (such as `Math.DivRem`) before it got merged.

I only relented on that with the premise that these would not be public until after the other work was completed. So if we are reverting this, then any new PR should likely just complete the additional work so we don't risk shipping something that isn't actually functioning the way users expect.

kunalspathak commented:
>  So if we are reverting this

I was just planning to do the following:
- add back the `DivRem` methods in `ref
- add `RequiresPreviewFeaturesAttribute` on the new APIs.
- revert the fake library workaround added to make test projects work

Do we agree that it is the best course or should we completely revert this PR, address the codegen issues and likely close https://github.com/dotnet/runtime/issues/82194? I am fine with either solution.

jkotas commented:
Yes, I think that it is the best course of action.

kunalspathak commented:
> Yes, I think that it is the best course of action.

https://github.com/dotnet/runtime/pull/82221

huoyaoyuan commented:
Glad to see this finally get merged!
